[New post] How to build a VTubing app with Amazon Interactive Video Service and VRoid

tonyv opened the pull request:
…eo Service and VRoid

# Content Pull Request Checklist

Please check to make sure your content submission meets the following requirements by adding an `x` between the brackets:

- [X] The title of this pull request reflects the title of my content.
- [X] I have reviewed the [Content Review Checklist](https://github.com/build-on-aws/content/blob/main/CONTENT_REVIEW_CHECKLIST.md).
- [X] This pull request contains one piece of content (post or page).

**Important: if you are submitting a PR to update existing content, please ensure you [sync'ed your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) first before you created your branch.**

Is this pull request linked to a Content Proposal or Idea Suggestion Issue? If so, please link to it below using the number or full link.

```text
Resolves #5

or

Content for https://github.com/build-on-aws/content/issues/5
```

**By submitting this pull request, I confirm I own this content, have permission to use each image included in this content, and that it will be released under the [CC BY-SA 4.0 SA License](https://github.com/build-on-aws/content/blob/main/LICENSE). Notwithstanding the above, if you provide source code text within your content piece, you are providing such source code under the [MIT-0](https://opensource.org/license/mit-0/) license.**

darko-mesaros commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,255 @@
+| 🧩 Prerequisites    | - [AWS Account](https://aws.amazon.com/resources/create-account/)<br>- [Pixiv Account](https://accounts.pixiv.net/signup) <br> |
+| 💻 Code Sample      | Code sample used in tutorial on [GitHub](#)                                                                                    |
+| 📢 Feedback         | <a href="https://pulse.buildon.aws/survey/DEM0H5VW" target="_blank">Any feedback, issues, or just a</a> 👍 / 👎 ?              |
+| ⏰ Last Updated     | YYYY-MM-DD                                                                                                                     |
```
Can you just update the date here.

darko-mesaros commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,255 @@
+| ⏱ Time to complete  | 15 minutes                                                                                                                     |
+| 💰 Cost to complete | Free when using the AWS Free Tier or USD 1.01                                                                                  |
+| 🧩 Prerequisites    | - [AWS Account](https://aws.amazon.com/resources/create-account/)<br>- [Pixiv Account](https://accounts.pixiv.net/signup) <br> |
+| 💻 Code Sample      | Code sample used in tutorial on [GitHub](#)                                                                                    |
```
Add the github link to the code repo here.

darko-mesaros commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,255 @@
+
+## Part 1 - Integrating with the VRoid Hub API and loading an avatar
+
+First, [sign up for a Pixiv account](https://accounts.pixiv.net/signup) and follow the [VRoid Hub quick start guide](https://developer.vroid.com/en/api/quick-start.html) to create an application. Once you create an application, you can reference the [code sample](https://github.com/pixiv/VRoidHub-API-Example) provided by Pixiv that demonstrates how to download a character on the VRoid Hub as a VR ...
```
I'd love to see some additional instructions here, as the VRoid quick start guide was a bit confusing (few items in the new application wizard I just made up). Just to make it a bit smoother to the reader, you should anticipate this confusion for people who have never used VRoid

darko-mesaros replied:
The code sample is leading us to a Repo that is mostly in Japanese 😓 I was not sure if I should read it and follow instructions. If this is not necessary to proceed with the tutorial I suggest you move it to a little Indented segment:
> Like this
And then continue with the steps needed - and that is to download the VRM file itself+

darko-mesaros commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
````diff
@@ -0,0 +1,255 @@
+<script src="https://cdn.jsdelivr.net/npm/kalidokit@1.1/dist/kalidokit.umd.js"></script>
+```
+
+Now let’s add code for our character rigging logic and create the animateVRM function. The animateVRM function will be used to get landmark data from MediaPipe Holistic passed in via the results argument. In the context of rigging, landmark data refers to specific points of a body part. Using the landmark data from MediaPipe Holistic, we can pass it to Kalidokit to animate specific body parts. Yo ...
````
Missing link to [here]

darko-mesaros commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
````diff
@@ -0,0 +1,255 @@
+
+After importing these libraries, let’s use the Three.js library to load a [scene object](https://threejs.org/docs/#api/en/scenes/Scene). A scene is used in Three.js to load in our 3D character or avatar. You will then use the [WebGLRenderer](https://threejs.org/docs/#api/en/renderers/WebGLRenderer) provided by Three.js to to actually display the scene. Using the renderer object created from WebGL ...
+
+```javascript
````
Where is this code supposed to go? The `index.html` file?

darko-mesaros commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,255 @@
+
+With that context set, I will walk you through a step by step process of building a web app to load a 3D character from VRoid Hub, animate it based on your body movements, and live stream to Amazon Interactive Video Service (IVS). [VRoid Hub](https://hub.vroid.com/en) is an online platform created by Pixiv used to host and share 3D character models. You will be using the VRoid Hub API to import a ...
+
+> The code for this tutorial is available on [Github](#). You can also try a [live demo](https://dev.d218eir1ybnzul.amplifyapp.com/) of this web app.
```
Missing the github link here.

darko-mesaros commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,255 @@
+
+## Part 3 - Animating a virtual character with your own body movements
+
+To start animating your virtual character, install the [Kalidokit](https://github.com/yeemachine/kalidokit) and [MediaPipe Holistic](https://github.com/google/mediapipe/blob/master/docs/solutions/holistic.md) libraries. MediaPipe Holistic is a computer vision pipeline used to track a user’s body movements, facial expressions, and hand gestures. This is useful for animating your digital avatar to  ...
```
You are using the term "Install" here but I am not sure if there is any installation to be made. If we are just importing them to the HTML there should be nothing to install. I'd suggest you change the word from "install" to "import" or something similar :)

darko-mesaros commented:
OVERALL FEEDBACK: I could not finish this tutorial, I believe critical parts of it are missing. Namely the JavaScript code that was presented, I am not really sure where to write that as adding it too the `index.html` file makes no sense at least in the way it's presented. 

The goal of this tutorial is to have any one be able to pick it up and wind up with the same results as you had! So, I'd love if you would be a bit more granular with the steps needed to get there. Thank you!

tonyv commented:
@darko-mesaros Very much appreciate the thorough review and feedback. I've addressed all feedback and just waiting for links to the Github repo referenced in the blog before pushing another commit.

tonyv commented:
Hi @darko-mesaros - Quick update that I'm still waiting on the Github repo links. Would you be able to review the other changes I have made in the meantime?

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
````diff
@@ -0,0 +1,540 @@
+</html>
+```
+
+## Part 3 - Rendering a virtual character with the three VRM SDK
````
```suggestion
## Part 3 - Rendering a Virtual Character With the Three VRM SDK
```

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,540 @@
+
+## Part 4 - Animating a virtual character with your own body movements
+
+To start animating your virtual character, add `<script>` elements for the [Kalidokit](https://github.com/yeemachine/kalidokit) library, [MediaPipe Holistic](https://github.com/google/mediapipe/blob/master/docs/solutions/holistic.md) library, and camera utility module from MediaPipe to the `<head>` element in `index.html`. MediaPipe Holistic is a computer vision pipeline used to track a user’s bo ...
```
```suggestion
To start animating your virtual character, add `<script>` elements for the [Kalidokit](https://github.com/yeemachine/kalidokit) library, [MediaPipe Holistic](https://github.com/google/mediapipe/blob/master/docs/solutions/holistic.md) library, and camera utility module from MediaPipe to the `<head>` element in `index.html`. MediaPipe Holistic is a computer vision pipeline used to track a user’s body movements, facial expressions, and hand gestures. This is useful for animating your digital avatar to mimic your own movements. Kalidokit includes the use of blendshapes for facial animation and kinematics solvers for body movements to create more realistic digital avatars. Blendshapes are a technique used in character animation to create a wide range of facial animations. Kinematics solvers are algorithms used to calculate the position and orientation of an avatar’s limbs. When making our avatar animate, aka character rigging, a kinematics solver helps determine how a character’s joints and bones should move to achieve a desired pose or animation. In short, MediaPipe Holistic tracks your physical movements while Kalidokit takes those as inputs to animate your avatar. The camera utility module from MediaPipe will simplify the process of providing our front-facing camera input to MediaPipe Holistic. MediaPipe Holistic needs this camera input to do hand, face and body movement tracking.
```

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
````diff
@@ -0,0 +1,540 @@
+camera.start();
+```
+
+## Part 5 - Live stream your virtual character to Amazon IVS
````
```suggestion
## Part 5 - Live Stream Your Virtual Character to Amazon IVS
```

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
````diff
@@ -0,0 +1,540 @@
+};
+```
+
+Once we have a MediaStream we want to publish to an audience, we need to join a stage. Joining a stage enables us to live stream the video feed to the audience or other participants in the stage. If we don’t want to live stream anymore, we can leave the stage. Let’s add event listeners that listen for click events when an end user clicks the join or leave stage buttons and implement the appropria ...
````
```suggestion
Once we have a MediaStream we want to publish to an audience and we need to join a stage. Joining a stage enables us to live stream the video feed to the audience or other participants in the stage. If we don’t want to live stream anymore, we can leave the stage. Let’s add event listeners that listen for click events when an end user clicks the join or leave stage buttons and implement the appropriate logic.
```

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
````diff
@@ -0,0 +1,540 @@
+};
+```
+
+Finally, let’s add some logic to listen for Stage events. These events occur when the state of the stage you’ve joined changes such as when someone joins or leaves it. Using these events, you can dynamically update the HTML code to display a new participant’s video feed when they join or remove it from display when they leave. The setupParticipant and teardownParticipant functions do each of thes ...
````
```suggestion
Finally, let’s add some logic to listen for Stage events. These events occur when the state of the stage you’ve joined changes such as when someone joins or leaves it. Using these events, you can dynamically update the HTML code to display a new participant’s video feed when they join or remove it from display when they leave. The setupParticipant and teardownParticipant functions do each of these actions respectively. As a step we call the join method on the stage object to join the stage.
```

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,540 @@
+## Conclusion
+
+In this tutorial, you created a VTubing app by leveraging Pixiv’s SDKs to display a virtual avatar character and live stream it using Amazon IVS. VTubing opens the doors to an exciting world of virtual content creation. By following the steps outlined in this tutorial, you have gained the knowledge and tools necessary to bring your unique virtual persona to life.
+
```
Great conclusion! We want to invite readers to stay in the blog repository longer. Please add a sentence or two that is apropos of your topic with 2-3 blog links.  

EX: "To read more/learn more about X, visit (LINK) and (LINK)."

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,540 @@
+---
+title: "How to build a VTubing app with Amazon Interactive Video Service and VRoid"
+description: "A step by step guide to set up a 3D virtual avatar that follows your body movements and leverage live streaming from Amazon IVS to broadcast to a large audience."
+tags:
```
We can only use 5 tags. Please choose the best ones.

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,540 @@
+---
+title: "How to build a VTubing app with Amazon Interactive Video Service and VRoid"
+description: "A step by step guide to set up a 3D virtual avatar that follows your body movements and leverage live streaming from Amazon IVS to broadcast to a large audience."
```
The descriptions need to be 60 characters or fewer. They can't repeat the messaging in the title or use the same words. They should focus on what solution the reader gets from reading the piece. I wrote an example below but feel free to rewrite. 

```suggestion
description: "Make a 3D avatar with live streaming to broadcast to a large audience."
```

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,540 @@
+---
+title: "How to build a VTubing app with Amazon Interactive Video Service and VRoid"
```
```suggestion
title: "How To Build a VTubing App With Amazon Interactive Video Service and VRoid"
```

DalM2023 commented on tutorials/build-a-vtubing-app-with-amazon-ivs-and-vroid/index.md:
```diff
@@ -0,0 +1,540 @@
+---
```
Hello, 


What a great opening line and what a fun project! 

Hello,

What a fun project! 

This is a well written blog with good structure and sources.  My edits focus primarily on correcting titles to AP Style and proper name capitalization errors. Please see my notes below. 

Overall, well done!

cherbk commented:
@damonk10 Please do final review but not publish. The author has requested this piece be published in mid  January. Thanks

tonyv commented:
> @damonk10 Please do final review but not publish. The author has requested this piece be published in mid January. Thanks

@damonk10 @cherbk Kindly following up on this. If possible, could this be published around Jan 15?
